### ISSUE TITLE
pyright does not distinguish between class and instance variables, resulting in wrong diagnostics

### ISSUE THREAD

**Issue by maxnoe** [microsoft/pyright#5225]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**

If a class variable is shadowed by an instance variable, pyright still uses the type of the class variable to provide type checks / diagnostics.

Interestingly enough, it proposes the correct things in the auto-complete suggestions.

This is a common situation with things like the descriptor protocol.

**To Reproduce**

Consider the much simplified code below from a common setup using descriptor classes to describe members of a data-holding class, common for ORMs or similar things:

```
class Field:
    pass


class Point:
    foo = Field()

    def __init__(self, foo):
        self.foo = foo


p = Point(foo={})
print(p.foo.items())
``` 

Pyright complains about the last line: "Cannot access member "items" for type "Field", checking the class variable instead of the instance variable

**Expected behavior**

Distinguish between class and instance variables.

**Screenshots or Code**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or imports all referenced symbols.

See above.

**VS Code extension or command-line**

neovim using lsp

**Additional context**
